pkg/progressreader/progressstatus.go

```diff
@@ -0,0 +1,78 @@
+package progressreader
+
+import (
+	"bytes"
+	"io"
+	"sync"
+)
+
+// ProgressStatus keeps track of one or more observers watching the progress
+// of an operation. For example, if multiple clients are trying to pull an
+// image, they share a ProgressStatus for the download operation.
+type ProgressStatus struct {
+	sync.Mutex
+	// c is a channel that observers block on, waiting for the operation
+	// to finish.
+	c chan struct{}
+	// observers is a list of io.Writers which receive a copy of the
+	// progress output.
+	observers []io.Writer
+	// history is a buffer of the progress output so far, so a new observer
+	// can catch up.
+	history bytes.Buffer
+}
+
+// NewProgressStatus returns a ProgressStatus structure
+func NewProgressStatus() *ProgressStatus {
+	return &ProgressStatus{
+		c:         make(chan struct{}),
+		observers: []io.Writer{},
+	}
+}
+
+// Write adds data to the history buffer, and also writes it to all current
+// observers.
+func (ps *ProgressStatus) Write(p []byte) (n int, err error) {
+	ps.Lock()
+	defer ps.Unlock()
+	ps.history.Write(p)
+	for _, w := range ps.observers {
```

Comment:
We need to hold the lock while iterating over `ps.observers`. It would be possible to make a copy of this slice and then release the lock, but I'm not sure blocking here is really harmful.